### PR TITLE
porter: support layer2

### PR DESCRIPTION
**Pull request by duanjiong**
Signed-off-by: Duan Jiong <djduanjiong@gmail.com>

### CODE REVIEW

File: pkg/layer2/announcer.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	portererror "github.com/kubesphere/porter/pkg/errors"
+	"github.com/kubesphere/porter/pkg/kubeutil"
+	"net"
+	"os"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"time"
+)
+
+// Announce is used to "announce" new IPs mapped to the node's MAC address.
```
Comment (zheng1) [inline pkg/layer2/announcer.go:14]: Please add a comment about `Referring to the code in https://github.com/metallb/metallb/tree/main/internal/layer2`

Comment (duanjiong) [reply pkg/layer2/announcer.go:14]: ok,  but there are many differences

**Comment by zheng1** [comment]
I think this PR is okay, can it be merged now?

File: pkg/layer2/announcer.go (hunk 1 above)
Comment (zheng1) [reply pkg/layer2/announcer.go:14]: OK, maybe this is better
```go
// Referenced part of the code in https://github.com/metallb/metallb/tree/main/internal/layer2
```

**Comment by duanjiong** [comment]
> I think this PR is okay, can it be merged now?

not ready, i need test more and fix some doc

**Comment by duanjiong** [comment]
#75 #31

File: pkg/ipam/ipam.go [hunk 2]
```diff
@@ -155,7 +165,11 @@ func (i *IPAM) useFinalizerIfNeeded(eip *networkv1alpha1.Eip) (bool, error) {
 }
 
 func (i *IPAM) AssignIP(serv *corev1.Service) (*AssignIPResponse, error) {
-	return i.ds.AssignIP(serv.Name, serv.Namespace)
+	lbType := serv.Annotations[constant.PorterLBTypeAnnotationKey]
+	if lbType == "" {
+		return nil, errors.NewEIPLBTypeNotFoundError()
```
Comment (zheng1) [inline pkg/ipam/ipam.go:170]: default type should be bgp

File: api/v1alpha1/eip_types.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ type EipSpec struct {
 	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:Pattern=`^([0-9]{1,3}\.){3}[0-9]{1,3}(\/([0-9]|[1-2][0-9]|3[0-2]))?$`
 	Address       string `json:"address,omitempty"`
+	LBType        string `json:"lbType,omitempty"`
```
Comment (zheng1) [inline api/v1alpha1/eip_types.go:28]: How about change `LBType` to `Protocol` like metalLB

Comment (duanjiong) [reply api/v1alpha1/eip_types.go:28]: I don't think it matters, maybe it's better to use different fields

File: pkg/ipam/ipam.go (hunk 2 above)
Comment (duanjiong) [reply pkg/ipam/ipam.go:170]: I don't think we need to make decisions for users

**Pull request closed without merging**